File: src/test/java/cube8540/oauth/authentication/oauth/client/application/OAuth2ClientInitializerTest.java
package cube8540.oauth.authentication.oauth.client.application;

import cube8540.oauth.authentication.oauth.client.domain.OAuth2Client;
import cube8540.oauth.authentication.oauth.client.domain.OAuth2ClientRepository;
import cube8540.oauth.authentication.security.AuthorityDetails;
import cube8540.oauth.authentication.security.AuthorityDetailsService;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.mockito.ArgumentCaptor;
import org.springframework.core.env.Environment;
import org.springframework.security.crypto.password.PasswordEncoder;

import java.util.Set;

import static cube8540.oauth.authentication.oauth.client.application.OAuth2ClientApplicationTestHelper.CLIENT_ID;
import static cube8540.oauth.authentication.oauth.client.application.OAuth2ClientApplicationTestHelper.ENCODING_SECRET;
import static cube8540.oauth.authentication.oauth.client.application.OAuth2ClientApplicationTestHelper.INITIALIZE_GRANT_TYPE;
import static cube8540.oauth.authentication.oauth.client.application.OAuth2ClientApplicationTestHelper.INITIALIZE_REDIRECT_URI;
import static cube8540.oauth.authentication.oauth.client.application.OAuth2ClientApplicationTestHelper.OWNER;
import static cube8540.oauth.authentication.oauth.client.application.OAuth2ClientApplicationTestHelper.RAW_CLIENT_ID;
import static cube8540.oauth.authentication.oauth.client.application.OAuth2ClientApplicationTestHelper.RAW_INITIALIZE_GRANT;
import static cube8540.oauth.authentication.oauth.client.application.OAuth2ClientApplicationTestHelper.RAW_INITIALIZE_REDIRECT_URI;
import static cube8540.oauth.authentication.oauth.client.application.OAuth2ClientApplicationTestHelper.RAW_OWNER;
import static cube8540.oauth.authentication.oauth.client.application.OAuth2ClientApplicationTestHelper.RAW_SCOPES;
import static cube8540.oauth.authentication.oauth.client.application.OAuth2ClientApplicationTestHelper.SCOPES;
import static cube8540.oauth.authentication.oauth.client.application.OAuth2ClientApplicationTestHelper.SECRET;
import static cube8540.oauth.authentication.oauth.client.application.OAuth2ClientApplicationTestHelper.makeAuthorityDetails;
import static cube8540.oauth.authentication.oauth.client.application.OAuth2ClientApplicationTestHelper.makeClientRepository;
import static cube8540.oauth.authentication.oauth.client.application.OAuth2ClientApplicationTestHelper.makeDefaultClient;
import static cube8540.oauth.authentication.oauth.client.application.OAuth2ClientApplicationTestHelper.makeEmptyClientRepository;
import static cube8540.oauth.authentication.oauth.client.application.OAuth2ClientApplicationTestHelper.makeEncoder;
import static cube8540.oauth.authentication.oauth.client.application.OAuth2ClientApplicationTestHelper.makeInitializerAuthorityDetailsService;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@DisplayName("클라이언트 초기화 테스트")
public class OAuth2ClientInitializerTest {
    private static final String USERNAME_KEY = "init-user.username";

    private static final String CLIENT_ID_KEY = "init-oauth-client.client-id";
    private static final String CLIENT_SECRET_KEY = "init-oauth-client.client-secret";
    private static final String CLIENT_NAME = "init-oauth-client.client-name";
    private static final String CLIENT_GRANT_TYPE_KEY = "init-oauth-client.client-grant-type";
    private static final String CLIENT_REDIRECT_URI = "init-oauth-client.client-redirect-uri";

    @Test
    @DisplayName("초기화 클라이언트가 저장소에 있을시")
    void initializeWhenClientIsExists() {
        OAuth2Client client = makeDefaultClient();
        OAuth2ClientRepository repository = makeClientRepository(CLIENT_ID, client);
        PasswordEncoder encoder = makeEncoder(SECRET, ENCODING_SECRET);
        Set<AuthorityDetails> authorityDetails = makeAuthorityDetails(RAW_SCOPES);
        AuthorityDetailsService authorityDetailsService = makeInitializerAuthorityDetailsService(authorityDetails);
        Environment environment = mock(Environment.class);
        OAuth2ClientInitializer initializer = new OAuth2ClientInitializer(repository, encoder, authorityDetailsService);

        when(environment.getRequiredProperty(CLIENT_ID_KEY)).thenReturn(RAW_CLIENT_ID);
        when(environment.getRequiredProperty(CLIENT_SECRET_KEY)).thenReturn(SECRET);
        when(environment.getRequiredProperty(CLIENT_GRANT_TYPE_KEY)).thenReturn(RAW_INITIALIZE_GRANT);
        when(environment.getRequiredProperty(CLIENT_REDIRECT_URI)).thenReturn(RAW_INITIALIZE_REDIRECT_URI);
        when(environment.getRequiredProperty(CLIENT_NAME)).thenReturn(CLIENT_NAME);
        when(environment.getRequiredProperty(USERNAME_KEY)).thenReturn(RAW_OWNER);

        initializer.initialize(environment);
        verify(repository, never()).save(any());
    }

    @Test
    @DisplayName("초기화 클라이언트가 저장소에 없을시")
    void initializeWhenClientNotExists() {
        ArgumentCaptor<OAuth2Client> clientCaptor = ArgumentCaptor.forClass(OAuth2Client.class);
        OAuth2ClientRepository repository = makeEmptyClientRepository();
        PasswordEncoder encoder = makeEncoder(SECRET, ENCODING_SECRET);
        Set<AuthorityDetails> authorityDetails = makeAuthorityDetails(RAW_SCOPES);
        AuthorityDetailsService authorityDetailsService = makeInitializerAuthorityDetailsService(authorityDetails);
        Environment environment = mock(Environment.class);
        OAuth2ClientInitializer initializer = new OAuth2ClientInitializer(repository, encoder, authorityDetailsService);

        when(environment.getRequiredProperty(CLIENT_ID_KEY)).thenReturn(RAW_CLIENT_ID);
        when(environment.getRequiredProperty(CLIENT_SECRET_KEY)).thenReturn(SECRET);
        when(environment.getRequiredProperty(CLIENT_GRANT_TYPE_KEY)).thenReturn(RAW_INITIALIZE_GRANT);
        when(environment.getRequiredProperty(CLIENT_REDIRECT_URI)).thenReturn(RAW_INITIALIZE_REDIRECT_URI);
        when(environment.getRequiredProperty(CLIENT_NAME)).thenReturn(CLIENT_NAME);
        when(environment.getRequiredProperty(USERNAME_KEY)).thenReturn(RAW_OWNER);

        initializer.initialize(environment);
        verify(repository, times(1)).save(clientCaptor.capture());
        assertEquals(CLIENT_ID, clientCaptor.getValue().getClientId());
        assertEquals(ENCODING_SECRET, clientCaptor.getValue().getSecret());
        assertEquals(CLIENT_NAME, clientCaptor.getValue().getClientName());
        assertEquals(INITIALIZE_GRANT_TYPE, clientCaptor.getValue().getGrantTypes());
        assertEquals(INITIALIZE_REDIRECT_URI, clientCaptor.getValue().getRedirectUris());
        assertEquals(OWNER, clientCaptor.getValue().getOwner());
        assertEquals(SCOPES, clientCaptor.getValue().getScopes());
    }
}

File: src/test/kotlin/cube8540/oauth/authentication/oauth/client/application/OAuth2ClientInitializerTest.kt
package cube8540.oauth.authentication.oauth.client.application

import cube8540.oauth.authentication.oauth.client.domain.ClientOwner
import cube8540.oauth.authentication.oauth.client.domain.OAuth2Client
import cube8540.oauth.authentication.oauth.client.domain.OAuth2ClientId
import cube8540.oauth.authentication.oauth.client.domain.OAuth2ClientRepository
import cube8540.oauth.authentication.security.AuthorityCode
import cube8540.oauth.authentication.security.AuthorityDetails
import cube8540.oauth.authentication.security.AuthorityDetailsService
import io.mockk.*
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Test
import org.springframework.core.env.Environment
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.security.oauth2.core.AuthorizationGrantType
import java.net.URI
import java.util.Optional

class OAuth2ClientInitializerTest {

    private val repository: OAuth2ClientRepository = mockk {
        every { save(any()) } returnsArgument 0
    }
    private val encoder: PasswordEncoder = mockk()
    private val authorityDetailsService: AuthorityDetailsService = mockk()

    private val initializer = OAuth2ClientInitializer(repository, encoder,  authorityDetailsService)

    @Test
    fun `initialize client is exists`() {
        val environment: Environment = mockk {
            every { getRequiredProperty(OAuth2ClientInitializer.CLIENT_ID_KEY) } returns "clientId"
            every { getRequiredProperty(OAuth2ClientInitializer.CLIENT_SECRET_KEY) } returns "clientSecret"
            every { getRequiredProperty(OAuth2ClientInitializer.CLIENT_GRANT_TYPE_KEY) } returns "client_credentials,authorization_code,password"
            every { getRequiredProperty(OAuth2ClientInitializer.CLIENT_REDIRECT_URI) } returns "http://localhost:8080/auth,http://localhost:8081"
            every { getRequiredProperty(OAuth2ClientInitializer.CLIENT_NAME) } returns "clientName"
            every { getRequiredProperty(OAuth2ClientInitializer.USERNAME_KEY) } returns "username"
        }
        val client: OAuth2Client = mockk()

        every { repository.findByClientId(OAuth2ClientId("clientId")) } returns Optional.of(client)

        initializer.initialize(environment)
        verify { repository.save(any()) wasNot Called }
    }

    @Test
    fun `initialize client is not exists`() {
        val environment: Environment = mockk {
            every { getRequiredProperty(OAuth2ClientInitializer.CLIENT_ID_KEY) } returns "clientId"
            every { getRequiredProperty(OAuth2ClientInitializer.CLIENT_SECRET_KEY) } returns "clientSecret"
            every { getRequiredProperty(OAuth2ClientInitializer.CLIENT_GRANT_TYPE_KEY) } returns "client_credentials,authorization_code,password"
            every { getRequiredProperty(OAuth2ClientInitializer.CLIENT_REDIRECT_URI) } returns "http://localhost:8080/auth,http://localhost:8081"
            every { getRequiredProperty(OAuth2ClientInitializer.CLIENT_NAME) } returns "clientName"
            every { getRequiredProperty(OAuth2ClientInitializer.USERNAME_KEY) } returns "username"
        }
        val initializeAuthorities: Set<AuthorityDetails> = setOf(
            mockk { every { code } returns "init-code-1" },
            mockk { every { code } returns "init-code-2" },
            mockk { every { code } returns "init-code-3" }
        )

        val clientCaptor = slot<OAuth2Client>()

        every { repository.findByClientId(OAuth2ClientId("clientId")) } returns Optional.empty()
        every { encoder.encode("clientSecret") } returns "encodedClientSecret"
        every { authorityDetailsService.loadInitializeAuthority() } returns initializeAuthorities
        every { repository.save(capture(clientCaptor)) } returnsArgument 0

        initializer.initialize(environment)
        assertThat(clientCaptor.captured.clientId).isEqualTo(OAuth2ClientId("clientId"))
        assertThat(clientCaptor.captured.secret).isEqualTo("encodedClientSecret")
        assertThat(clientCaptor.captured.clientName).isEqualTo("clientName")
        assertThat(clientCaptor.captured.grantTypes).isEqualTo(setOf(AuthorizationGrantType.CLIENT_CREDENTIALS, AuthorizationGrantType.AUTHORIZATION_CODE, AuthorizationGrantType.PASSWORD))
        assertThat(clientCaptor.captured.scopes).isEqualTo(setOf(AuthorityCode("init-code-1"), AuthorityCode("init-code-2"), AuthorityCode("init-code-3")))
        assertThat(clientCaptor.captured.redirectUris).isEqualTo(setOf(URI.create("http://localhost:8080/auth"), URI.create("http://localhost:8081")))
        assertThat(clientCaptor.captured.owner).isEqualTo(ClientOwner("username"))
    }

    @AfterEach
    fun clear() {
        clearAllMocks()
    }
}